client/src/main/java/io/pravega/client/stream/impl/SegmentSelector.java

```diff
@@ -126,14 +130,22 @@ public Segment getSegmentForEvent(String routingKey) {
     }
 
     @Synchronized
-    private List<PendingEvent> updateSegmentsUponSealed(StreamSegments newStreamSegments, Segment sealedSegment,
+    private List<PendingEvent> updateSegmentsUponSealed(StreamSegmentsWithPredecessors successors, Segment sealedSegment,
                                                         Consumer<Segment> segmentSealedCallback) {
-        currentSegments = newStreamSegments;
-        createMissingWriters(segmentSealedCallback, newStreamSegments.getDelegationToken());
-        log.debug("Fetch unacked events for segment: {}, and adding new segments {}", sealedSegment, newStreamSegments);
-        List<PendingEvent> toResend = writers.get(sealedSegment).getUnackedEventsOnSeal();
-        writers.remove(sealedSegment); //remove this sealed segment writer.
-        return toResend;
+        if (successors == null) {
+            //stream is deleted, hence successors is null.
+            writers.get(sealedSegment).getUnackedEventsOnSeal()
+                    .forEach(pendingEvent -> pendingEvent.getAckFuture()
+                            .completeExceptionally(new NoSuchSegmentException(sealedSegment.toString())));
```

Comment:
This is really dangerous. Never complete a Future while holding a lock. The Future's downstream handlers may execute synchronously on this thread while holding your lock, which can quickly lead to a deadlock.

First collect your targets while holding the lock, then complete those futures outside of the lock.

Reply:
Thank for pointing it out. Updated the code to cancel all the futures on an executor.